### DISCUSSION TITLE
[Coroutine] Create a global coroutine that doesn't stop after one step

### DISCUSSION THREAD

**Discussion by PoppyClick** [cocotb/cocotb#4237] [Q&A]
Hello,

First, this is not an issue with Cocotb but with my understanding of the tool, hence the discussion.

I'm trying to create a global timeout coroutine that would fail any test after reaching a certain simulation time.
After a lot of headaches, I've realised that the timeout function didn't work after leaving a test.

To narrow down the problem, I've written the following example :
```python
async def timeout(time):
  await Timer(time, "ns")
  assert False

@cocotb.test
async def tst_1(dut):
  cocotb.start_soon(timeout(100))
  # Wait half of the timeout
  await Timer(50, "ns")
  # Then move to next step

@cocotb.test
async def tst_2(dut):
  await Timer(100, "ns")
  print("No assert received")

@cocotb.test
async def tst_3(dut):
  cocotb.start_soon(timeout(100))
  # Wait half of the timeout
  await Timer(150, "ns")
  print("Assert received")
```
Which gives me the following log:
```log
#      0.00ns INFO     cocotb                             Running on ModelSim for Questa-64 version 2022.3 2022.07
# 
#      0.00ns INFO     cocotb                             Running tests with cocotb v1.9.1
# 
#      0.00ns INFO     cocotb                             Seeding Python random module with 1729870033
# 
#      0.00ns INFO     gpi                                Nothing visible via VPI
# 
#      0.00ns INFO     cocotb.regression                  Found test example.Simulations.test_000.test_000.tst_1
# 
#      0.00ns INFO     cocotb.regression                  Found test example.Simulations.test_000.test_000.tst_2
# 
#      0.00ns INFO     cocotb.regression                  Found test example.Simulations.test_000.test_000.tst_3
# 
#      0.00ns INFO     cocotb.regression                  running tst_1 (1/3)
# 
#     51.00ns INFO     cocotb.regression                  tst_1 passed
# 
#     51.00ns INFO     cocotb.regression                  running tst_2 (2/3)
# 
# No assert received
# 
#    152.00ns INFO     cocotb.regression                  tst_2 passed
# 
#    152.00ns INFO     cocotb.regression                  running tst_3 (3/3)
# 
#    252.00ns INFO     cocotb.Task 1.timeout              Test stopped by this forked coroutine
# 
#    253.00ns INFO     cocotb.regression                  tst_3 failed
# 
#                                                         Traceback (most recent call last):
# 
#                                                           File "test.py", line 22, in timeout
# 
#                                                             assert False
# 
#                                                         AssertionError: assert False
```

How could I better implement the global timeout behaviour ?
Thanks

**Comment by ktbarrett** [comment]
At the end of the test all Tasks are cancelled. You could also use `timeout_time` and `timeout_unit` argument to the `cocotb.test` decorator.

```python
@cocotb.test(timeout_time=100, timeout_unit='ns')
async def my_test(dut):
    ...
```

**Reply by PoppyClick** [reply]
That seemed to be the correct solution to my issue at first glance.
But it seems like it's a per-test timeout, while I was aiming for a global timeout where the simulation ends when reaching a certain time.

I've tried assigning the timeout value dynamically and readjusting it between each test, but cocotb seems to register the values before starting the simulation and doesn't update it afterward. Which makes sense. 

Something like a "global_task", or a "don't cancel task" option would solve my issue.

**Reply by ktbarrett** [reply]
Unfortunately that would break test reproducibility, so that is a feature we would probably not introduce.

If you expect that the tests are always run together in a particular order, it might be a good idea to combine the tests into one test.
